File: algorithms/data_structures/linked_list.py
"""
=====================================================================
Singly Linked List / Односвязный список
---------------------------------------------------------------------
prepend O(1), append O(n), reverse O(n).
=====================================================================
"""


class ListNode:
    def __init__(self, value, next=None):
        self.value = value
        self.next = next


class LinkedList:
    def __init__(self):
        self.head = None
        self.length = 0

    def prepend(self, value):
        self.head = ListNode(value, self.head)
        self.length += 1

    def append(self, value):
        node = ListNode(value)
        if not self.head:
            self.head = node
        else:
            curr = self.head
            while curr.next:
                curr = curr.next
            curr.next = node
        self.length += 1

    def insert_at(self, index: int, value):
        if not 0 <= index <= self.length:
            raise IndexError("Out of bounds")
        if index == 0:
            return self.prepend(value)

        prev = self.head
        for _ in range(index - 1):
            prev = prev.next
        prev.next = ListNode(value, prev.next)
        self.length += 1

    def remove_at(self, index: int):
        if not 0 <= index < self.length:
            raise IndexError("Out of bounds")

        if index == 0:
            removed = self.head
            self.head = self.head.next
        else:
            prev = self.head
            for _ in range(index - 1):
                prev = prev.next
            removed = prev.next
            prev.next = removed.next
        self.length -= 1
        return removed.value

    def reverse(self):
        """In-place разворот. Классика собеса."""
        prev = None
        curr = self.head
        while curr:
            curr.next, prev, curr = prev, curr, curr.next
        self.head = prev

    def to_list(self):
        result = []
        curr = self.head
        while curr:
            result.append(curr.value)
            curr = curr.next
        return result


if __name__ == "__main__":
    ll = LinkedList()
    ll.append(1); ll.append(2); ll.append(3)
    ll.prepend(0)
    print(ll.to_list())  # [0, 1, 2, 3]
    ll.reverse()
    print(ll.to_list())  # [3, 2, 1, 0]

File: algorithms/data_structures/linked_list.js
// =====================================================================
// Singly Linked List / Односвязный список
// ---------------------------------------------------------------------
// Узел: { value, next }
// Операции:
//   prepend       — O(1)
//   append        — O(n) без tail, O(1) если хранить tail
//   insertAt(k)   — O(n)
//   removeAt(k)   — O(n)
//   reverse       — O(n)
// Применение: реализация стека/очереди, LRU cache, hash table chaining.
// =====================================================================

class ListNode {
    constructor(value, next = null) {
        this.value = value;
        this.next = next;
    }
}

class LinkedList {
    constructor() {
        this.head = null;
        this.length = 0;
    }

    prepend(value) {
        this.head = new ListNode(value, this.head);
        this.length++;
    }

    append(value) {
        const node = new ListNode(value);
        if (!this.head) {
            this.head = node;
        } else {
            let curr = this.head;
            while (curr.next) curr = curr.next;
            curr.next = node;
        }
        this.length++;
    }

    insertAt(index, value) {
        if (index < 0 || index > this.length)
            throw new RangeError('Out of bounds');
        if (index === 0) return this.prepend(value);

        let prev = this.head;
        for (let i = 0; i < index - 1; i++) prev = prev.next;
        prev.next = new ListNode(value, prev.next);
        this.length++;
    }

    removeAt(index) {
        if (index < 0 || index >= this.length)
            throw new RangeError('Out of bounds');

        let removed;
        if (index === 0) {
            removed = this.head;
            this.head = this.head.next;
        } else {
            let prev = this.head;
            for (let i = 0; i < index - 1; i++) prev = prev.next;
            removed = prev.next;
            prev.next = removed.next;
        }
        this.length--;
        return removed.value;
    }

    // Классическая задача — разворот односвязного списка in-place
    reverse() {
        let prev = null;
        let curr = this.head;
        while (curr) {
            const next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        this.head = prev;
    }

    toArray() {
        const result = [];
        let curr = this.head;
        while (curr) {
            result.push(curr.value);
            curr = curr.next;
        }
        return result;
    }
}

const list = new LinkedList();
list.append(1);
list.append(2);
list.append(3);
list.prepend(0);
console.log(list.toArray()); // [0, 1, 2, 3]
list.reverse();
console.log(list.toArray()); // [3, 2, 1, 0]

module.exports = { ListNode, LinkedList };
